### DISCUSSION TITLE
Map Compare

### DISCUSSION THREAD

**Discussion by rullymartanto** [visgl/deck.gl#7627] [Q&A] [answered]
Do it possible to create map comparison same with mapbox compare?

https://docs.mapbox.com/mapbox-gl-js/example/mapbox-gl-compare/

**Comment by Pessimistress** [comment, marked as answer]
You can use the [multiple views](https://deck.gl/docs/developer-guide/views#using-multiple-views) feature.

See example [here](https://codepen.io/Pessimistress/pen/jOpoeLg)

**Reply by neelduttahere** [reply]
Hey @Pessimistress ,
I tried your approach, but at `getViews()`, since the default argument is 50, DeckGL seems to immediately re-render each time Props are updated from `updateSplitLocation`, but the updated `leftPercentage` is not accounted  for, since it is re-rendering when used as ReactComponent. But every time the dragging interaction rests, the split point gets reset to 50. Refer to the video below:

https://github.com/visgl/deck.gl/assets/59992993/3df1985e-f1c5-445c-ba06-a465432504e9



As you can see in the video, resting the slider brings back the split point to 50. Is there a better way to implement split mode in deck? Specially considering when we want the `<Map/>` component of `react-map-gl` to serve as the basemap in the first place. Due to this, I am forced to use base maps as `<TileLayer/>` for each `MapView`. Here is my implementation below:
```js
useEffect(() => {
    new Draggable(document.getElementById("handle"), {
        limit: {
          y: [0, 0],
        },
        onDrag: (_, x) => updateSplitLocation(x),
      });
  }, [deckRef]);

function updateSplitLocation(x) {
    const lPercentage = (x * 100) / (window.innerWidth - 48);
   // removing 48 pixels more to innerWidth because to compensate the width of nav sidebar in the app

// Calculate new view state
    const viewport = deckRef.current.deck.getViewports()[0];
    const newCenter = viewport.unproject([x, viewport.height / 2]);

    deckRef.current.deck.setProps({
      viewState: {
        longitude: newCenter[0],
        latitude: newCenter[1],
        zoom: viewport.zoom,
        pitch: viewport.pitch,
        bearing: viewport.bearing,
      },
      views: getViews(lPercentage),
    });
  }

function getViews(leftPercentage = 50) {
    return [
      new MapView({
        id: "left-side",
        x: 0,
        width: `${leftPercentage}%`,
        padding: { left: "100%" },
        controller: true,
      }),
      new MapView({
        id: "right-side",
        x: `${leftPercentage}%`,
        width: `${100 - leftPercentage}%`,
        padding: { right: "100%" },
        controller: true,
      }),
    ];
  }

return (
<>
   <DeckGL
        ref={deckRef}
        viewState={{
          longitude: MapCenter[1], // -> from redux state
          latitude: MapCenter[0], //-> from redux state
          zoom: zoom,
          pitch: 0,
          bearing: 0,
        }}
        onViewStateChange={(event) => {

          let viewState = event.viewState;
          deckRef.current.deck.setProps({ viewState });
          //optional code to update map center and zoom in redux
          //dispatch(setCenter([viewState.latitude, viewState.longitude]));
          //dispatch(setZoom(viewState.zoom));
        }}
        layers={splitLayers} //--> contains layers to be rendered programmatically
        layerFilter={({ layer, viewport }) => layer.id.startsWith(viewport.id)}
        views={getViews()}
      />
     <Box
          id="handle"
          css={{
            position: "absolute",
            zIndex: 9,
            marginLeft: "50%",
            height: "100vh",
            display: "flex",
          }}
        >
          <Box
            css={{
              background: "$defaultTone",
              boxShadow: `0 2px 8px rgba(0, 0, 0, 0.4)`,
              border: "1px solid $pixxelFontBright",
              width: 6,
              cursor: "col-resize",
              height: "100%",
              "&:hover": {
                border: "1px solid $accent",
              },
            }}
          />
        </Box>
</>
)
```

**Reply by Pessimistress** [reply]
If you are using React, then do things the React way. Do not call setProps. Save your view parameters with useState.

**Reply by neelduttahere** [reply]
Okay, that makes sense, now using useState to save and update viewport params. But I am still using setProps to update the views using the `getViews` function. The results are almost there (video attached below), but the map viewport seems to be shifting exponentially as I move the slider around, in the direction slider moves. Here is the code snippet below for the changes:

```js
function updateSplitLocation(x) {
    const lPercentage = (x * 100) / (window.innerWidth - 48);

    // Calculate new view state
    const viewport = deckRef.current.deck.getViewports()[0];
    const newCenter = viewport.unproject([x, viewport.height / 2]);

    setViewState({
      longitude: newCenter[0],
      latitude: newCenter[1],
      zoom: viewport.zoom,
      pitch: viewport.pitch,
      bearing: viewport.bearing,
    });

    deckRef.current.deck.setProps({
      views: getViews(lPercentage),
    });
  }
```


https://github.com/visgl/deck.gl/assets/59992993/f9bafc28-24da-4a31-a507-0321c688a1d8

**Reply by neelduttahere** [reply]
I finally figured it out. I will share a full example soon. Maybe we should add this as an example in DeckGL examples page, as this is a very popular use case for web-GIS-related features.

**Reply by Pessimistress** [reply]
> But I am still using setProps to update the views using the getViews function.

Your problem is here:

```jsx
<DeckGL
        ...
        views={getViews()} // <-- calling getViews with default leftPercentage=50
      />
```

**Reply by Tomas-Aguinaga** [reply]
@neelduttahere let me know please the example! This is a super common case as you say and i can create pr for the examples page
